### PR TITLE
Refactor Config mojo.

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/validation/SchemaValidator.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                    mapper.valueToTree(Collections.singletonMap(name, stringToObject(type, value))));
+            if (reports.isSuccess()) {
+                return null;
+            }
+
+            final List<String> errors = new ArrayList<>();
+            for (final ProcessingMessage pm : reports) {
+                errors.add(pm.getMessage());
+            }
+            if (errors.size() == 1) {
```
Comment (Flanker32) [inline azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/validation/SchemaValidator.java]: Change to `String.join("\n", errors)`?

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/ConfigMojo.java [hunk 2]
```diff
@@ -47,201 +53,220 @@
...
+            final List<MavenProject> configuredProjects = new ArrayList<>();
+            for (final MavenProject proj : allProjects) {
+                if (isProjectConfigured(proj)) {
+                    configuredProjects.add(proj);
+                } else {
+                    targetProjects.add(proj);
+                }
             }
-            return;
+            this.wrapper.putCommonVariable("projects", targetProjects);
```
Comment (Flanker32) [inline azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/ConfigMojo.java]: Dup with last line, remove this one?

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/ConfigMojo.java [hunk 3]
```diff
@@ -298,64 +315,63 @@ private void saveConfigurationToProject(MavenProject proj, boolean includeDeploy
...
-                    new ArrayList<MavenProject>());
+            this.appSettings.setIsPublic(this.wrapper.handle("configure-public", false));
         }
+
     }
 
-    private void configureAppName() throws IOException {
-        appName = prompt.promoteString("app name", "${project.artifactId}", "[A-Za-z0-9_\\.]+", true);
+    private void configureAppName() throws IOException, ExpressionEvaluationException, InvalidConfigurationException {
+        if (StringUtils.isNotBlank(appName)) {
```
Comment (Flanker32) [inline azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/ConfigMojo.java]: Is there any cases that config will run with `appName` not blank?

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/ConfigMojo.java (hunk 2 above)
Comment (andxu) [reply azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/ConfigMojo.java]: No, the first time, set all available projects for rendering the project list for user to select, on the second step, the target projects will be the project list selected by user

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/ConfigMojo.java (hunk 3 above)
Comment (andxu) [reply azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/ConfigMojo.java]: -DappName

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+            }
+            return null;
+        }
+        final boolean autoSelect = TemplateUtils.evalBoolean("auto_select", variables);
+        if (options.size() == 1) {
+            if (autoSelect) {
+                log.info(TemplateUtils.evalText("message.auto_select", variables));
+                return options.get(0);
+            }
+            if (!this.prompt.promoteYesNo(TemplateUtils.evalText("promote.one", variables),
```
Comment (Flanker32) [inline azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java]: Why not just select the one if the option is required?

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        final String type = (String) schema.get("type");
+
+        if (defaultValueCli != null) {
+            // valid against the property from cli parameter, if it passes, then we skip the configuration
+            final String errorMessage = validator.validateSchema(resourceName, propertyName, defaultValueCli.toString());
+            if (errorMessage == null) {
+                return defaultValueCli.toString();
+            }
+            System.out.println(TextUtils
+                    .yellow(String.format("Input validation failure for %s[%s[: ", propertyName, defaultValueCli.toString(), errorMessage)));
```
Comment (Flanker32) [inline azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java]: ` %s[%s]`?

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+            }
+        }
+        return selectedEntities;
+    }
+
+    public String handle(String templateId, boolean autoApplyDefault) throws InvalidConfigurationException, IOException, ExpressionEvaluationException {
+        return handle(templateId, autoApplyDefault, null);
+    }
+
+    public String handle(String templateId, boolean autoApplyDefault, Object defaultValueCli)
```
Comment (Flanker32) [inline azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java]: defaultValueCli is a bit confused, shall we change it to `cliParameter` or some word more clearly?

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    private String evaluateMavenExpression(String input) throws ExpressionEvaluationException {
+        if (input != null && input.contains("${")) {
+            return (String) expressionEvaluator.evaluate(input);
+        }
+        return input;
+    }
+
+    private Map<String, Object> mergeCommonProperties(Map<String, Object> map) {
```
Comment (Flanker32) [inline azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java]: `map.putAll(commonVariables);`?

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java (hunk 4 above)
Comment (andxu) [reply azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java]: If autoSelect is false, then we cannot take the only one choice.

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java (hunk 5 above)
Comment (andxu) [reply azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java]: fixed

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java (hunk 6 above)
Comment (andxu) [reply azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java]: fixed

File: azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java (hunk 7 above)
Comment (andxu) [reply azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/PromptWrapper.java]: fixed

**Pull request merged**